internal/generate/clusterserviceversion/clusterserviceversion_test.go

```diff
@@ -0,0 +1,379 @@
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clusterserviceversion
+
+import (
+	"bytes"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/format"
+
+	"github.com/blang/semver"
+	operatorversion "github.com/operator-framework/api/pkg/lib/version"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
+	"github.com/operator-framework/operator-registry/pkg/lib/bundle"
+	appsv1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/kubebuilder/pkg/model/config"
+	"sigs.k8s.io/yaml"
+
+	"github.com/operator-framework/operator-sdk/internal/generate/collector"
+	genutil "github.com/operator-framework/operator-sdk/internal/generate/internal"
+	kbutil "github.com/operator-framework/operator-sdk/internal/util/kubebuilder"
+	"github.com/operator-framework/operator-sdk/internal/util/projutil"
+)
+
+var (
+	testDataDir           = filepath.Join("..", "testdata")
+	csvDir                = filepath.Join(testDataDir, "clusterserviceversions")
+	csvBasesDir           = filepath.Join(csvDir, "bases")
+	csvNewLayoutBundleDir = filepath.Join(csvDir, "newlayout", "manifests")
+
+	// TODO: create a new testdata dir (top level?) that has both a "config"
+	// dir and a "deploy" dir that contains `kustomize build config/default`
+	// output to simulate actual manifest collection behavior. Using "config"
+	// directly is not standard behavior.
+	goTestDataDir     = filepath.Join(testDataDir, "non-standard-layout")
+	goAPIsDir         = filepath.Join(goTestDataDir, "api")
+	goManifestRootDir = filepath.Join(goTestDataDir, "config")
+	goCRDsDir         = filepath.Join(goManifestRootDir, "crds")
+)
+
+var (
+	col *collector.Manifests
+	cfg *config.Config
+)
+
+var (
+	baseCSV, baseCSVUIMeta, newCSV          *v1alpha1.ClusterServiceVersion
+	baseCSVStr, baseCSVUIMetaStr, newCSVStr string
+)
+
+var _ = BeforeSuite(func() {
+	col = &collector.Manifests{}
+	Expect(col.UpdateFromDirs(goManifestRootDir, goCRDsDir)).ToNot(HaveOccurred())
+
+	cfg = readConfigHelper(goTestDataDir)
+
+	initTestCSVsHelper()
+})
+
+var _ = Describe("Generating a ClusterServiceVersion", func() {
+	format.TruncatedDiff = true
+	format.UseStringerRepresentation = true
+
+	var (
+		g            Generator
+		buf          *bytes.Buffer
+		operatorName = "memcached-operator"
+		operatorType = projutil.OperatorTypeGo
+		version      = "0.0.1"
+	)
+
+	BeforeEach(func() {
+		buf = &bytes.Buffer{}
+	})
+
+	Describe("for the new Go project layout", func() {
+
+		Context("with correct Options", func() {
+
+			var (
+				tmp string
+				err error
+			)
+
+			BeforeEach(func() {
+				tmp, err = ioutil.TempDir(".", "")
+				ExpectWithOffset(1, err).ToNot(HaveOccurred())
```

Comment:
Why are we using `ExpectWithOffset()`?
Does `Expect(err).ToNot(HaveOccurred())` not work here?

Reply:
```
Unlike `Expect` and `Ω`, `ExpectWithOffset` takes an additional integer argument that is used to modify the call-stack offset when computing line numbers.

This is most useful in helper functions that make assertions. If you want Gomega's error message to refer to the calling line in the test (as opposed to the line in the helper function) set the first argument of `ExpectWithOffset` appropriately.
```

Seems like this is to bubble up error to the line number of the caller and not the helper itself.

Reply:
Exactly, and I've named all test helper functions that use gomega `Expect*()` functions with the `Helper` suffix.